pkg/operator/starter.go

```diff
@@ -45,53 +50,90 @@ const (
+	if guestKubeConfigFile != "" {
+		guestKubeConfig, err = client.GetKubeConfigOrInClusterConfig(guestKubeConfigFile, nil)
+		if err != nil {
+			return fmt.Errorf("Failed to use guest kubeconfig %s: %s", guestKubeConfigFile, err)
+		}
+		controlPlaneNamespace = controllerConfig.OperatorNamespace
+		guestKubeClient = kubernetes.NewForConfigOrDie(rest.AddUserAgent(guestKubeConfig, targetName))
+	}
 
-	configClient, err := configclient.NewForConfig(rest.AddUserAgent(controllerConfig.KubeConfig, targetName))
+	controlPlaneInformersForNamespaces := v1helpers.NewKubeInformersForNamespaces(controlPlaneKubeClient, "", controlPlaneNamespace)
+	guestKubeInformersForNamespaces := v1helpers.NewKubeInformersForNamespaces(guestKubeClient, "", defaultOperandNamespace)
+
+	// config.openshift.io client - use the guest cluster (Infrastructure)
+	guestConfigClient, err := configclient.NewForConfig(rest.AddUserAgent(guestKubeConfig, targetName))
 	if err != nil {
 		return err
 	}
-	configInformers := configinformer.NewSharedInformerFactoryWithOptions(configClient, resync)
+	guestConfigInformers := configinformer.NewSharedInformerFactoryWithOptions(guestConfigClient, resync)
 
-	apiExtClient, err := apiextclient.NewForConfig(rest.AddUserAgent(controllerConfig.KubeConfig, targetName))
+	// CRD client - use the guest cluster
+	guestAPIExtClient, err := apiextclient.NewForConfig(rest.AddUserAgent(guestKubeConfig, targetName))
 	if err != nil {
 		return err
 	}
 
-	// Create GenericOperatorclient. This is used by the library-go controllers created down below
+	// Create GenericOperatorClient. This is used by the library-go controllers created down below
+	// Operator CR is in the guest cluster.
 	gvr := operatorv1.SchemeGroupVersion.WithResource("csisnapshotcontrollers")
-	operatorClient, dynamicInformers, err := goc.NewClusterScopedOperatorClientWithConfigName(controllerConfig.KubeConfig, gvr, "cluster")
+	guestOperatorClient, dynamicInformers, err := goc.NewClusterScopedOperatorClientWithConfigName(guestKubeConfig, gvr, "cluster")
 	if err != nil {
 		return err
 	}
 
 	versionGetter := status.NewVersionGetter()
 
-	staticResourcesController := staticresourcecontroller.NewStaticResourceController(
-		"CSISnapshotStaticResourceController",
-		assets.ReadFile,
+	namespacedAssetFunc := namespaceReplacer(assets.ReadFile, "${CONTROLPLANE_NAMESPACE}", controlPlaneNamespace)
+	guestStaticResourceController := staticresourcecontroller.NewStaticResourceController(
+		"CSISnapshotGuestStaticResourceController",
+		namespacedAssetFunc,
 		[]string{
 			"volumesnapshots.yaml",
 			"volumesnapshotcontents.yaml",
 			"volumesnapshotclasses.yaml",
-			"webhook_config.yaml",
 		},
-		resourceapply.NewKubeClientHolder(kubeClient).WithAPIExtensionsClient(apiExtClient),
-		operatorClient,
+		resourceapply.NewKubeClientHolder(guestKubeClient).WithAPIExtensionsClient(guestAPIExtClient),
+		guestOperatorClient,
+		controllerConfig.EventRecorder,
```

Comment:
This recorder sends events to the mgtm cluster, we need a recorder for the guest cluster? https://github.com/openshift/aws-ebs-csi-driver-operator/pull/159/files#diff-0d623dfd885adb20f991bda4c2453aebd732ca6dbb4d1d4be6e79805c3b48de6R100-R105

Reply:
That's a good question, IMO all operator events should be sent to a single place, regardless what part of the operator is sending the event. Asking on slack.

Reply:
Added a patch that moves **all** events to the guest cluster. They will be created in openshift-cluster-storage-operator namespace and owned by the operator's Deployment, which may be in the mgmt cluster (and in a different namespace)